src/main/flight/alt_hold.c

```diff
@@ -0,0 +1,197 @@
+/*
+ * This file is part of Betaflight.
+ *
+ * Betaflight is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Betaflight is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Betaflight. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "platform.h"
+#include "alt_hold.h"
+
+#ifdef USE_ALTHOLD_MODE
+
+#include "flight/position.h"
+#include "flight/imu.h"
+#include "sensors/acceleration.h"
+#include "sensors/barometer.h"
+#include "config/config.h"
+#include "fc/runtime_config.h"
+#include "osd/osd.h"
+#include "common/printf.h"
+#include "common/maths.h"
+#include "math.h"
+#include "build/debug.h"
+
+
+PG_REGISTER_WITH_RESET_TEMPLATE(altholdConfig_t, altholdConfig, PG_ALTHOLD_CONFIG, 1);
+
+PG_RESET_TEMPLATE(altholdConfig_t, altholdConfig,
+    .velPidP = 30,
+    .velPidD = 0,
+
+    .altPidP = 75,
+    .altPidI = 20,
+
+    .minThrottle = 6,
+    .maxThrottle = 65,
+);
+
+
+typedef struct {
+    float max;
+    float min;
+    float kp;
+    float kd;
+    float ki;
+    float lastErr;
+    float integral;
+} simplePid_s;
+
+void simplePidInit(simplePid_s* simplePid, float min, float max, float kp, float kd, float ki)
+{
+    simplePid->max = max;
+    simplePid->min = min;
+    simplePid->kp = kp;
+    simplePid->kd = kd;
+    simplePid->ki = ki;
+    simplePid->lastErr = 0;
+    simplePid->integral = 0;
+}
+
+float simplePidCalculate(simplePid_s* simplePid, float dt, float targetValue, float measuredValue)
+{
+    float error = targetValue - measuredValue;
+
+    float pOut = simplePid->kp * error;
+
+    float iOut = simplePid->ki * simplePid->integral;
+
+    simplePid->integral += error * dt;
+
+    float derivative = (error - simplePid->lastErr) / dt;
+    float dOut = simplePid->kd * derivative;
+
+    float output = pOut + iOut + dOut;
+    output = constrainf(output, simplePid->min, simplePid->max);
+
+    simplePid->lastErr = error;
+    return output;
+}
+
+typedef struct {
+    simplePid_s altPid;
+    simplePid_s velPid;
+    float throttle;
+    float targetAltitude;
+    float measuredAltitude;
+    float measuredAccel;
+    float velocityEstimationAccel;  // based on acceleration
+    float startVelocityEstimationAccel;
```

Comment:
Is this really necessary? IMO this is only masking problem and proper handling of vertical velocity is beter approach.

Reply:
It's necessary.
 - We don't know acceleration drift. So we don't know a velocity drift.
 - We can take a very very conservative values, but in this case we lose a lot of useful data. Simulation confirms that using more available data makes flight more stable.
 - The best results are achieved when we allow velocity to have some really big drifts.
 - Continuous velocity drift compensated with I-term of the ALT pid, and it works really good.
 - But we should prevent initial velocity jump from zero velocity to our estimated drifted velocity.

> this is only masking problem and proper handling of vertical velocity is beter approach

You can suggest a better approach. Simple decrease of the velocity integrator coefficient to reduce velocity leak is not an option.

Reply:
( Did I mention Kalman before ;-) )

With current approach, you sample velocity with random noise + actual vertical speed. Control loop will compensate it eventually, but it takes some time. And it makes tuning a bit more difficult, control loop is compensating multiple error sources at the same time.

'Correct' way to handle this is to model accelerometer (offset) error and compensate it.
Current leak calculation is roughly equivalent to P regulator with velocity setpoint of zero. You may try PI regulator in this place (accumulating velocity error to compensate ACC offset).
Or you can use high-pass filter on accelerometer data (with very long time constant), thus removing DC offset and some drift. But velocity will have offset (direct result of integration), which must be compensated too .. 

Most approaches will lead to similar math, just derivation is different.

Also, don't trust simulation too much. It's quite easy to optimize for some simulation artefact instead of real problem (PID tuning to get noise reduction(nice graph) instead of dynamic response is one example)

Reply:
> But velocity will have offset (direct result of integration), which must be compensated too ..
Velocity drift is the only real issue here. AFAIK Kalman won't help with this, it only reduce noise but not a drift. And noise is not an issue as we don't use raw measurements but only their sum to get velocity.

> Also, don't trust simulation too much
I double checked this part with simulator and on real, the more drift we allow - the more stability we get in return. Actually I started with much lower CF, there was much less velocity drift but it had much worse stability.

`startVelocityEstimationAccel` - it handle the problem that we start velocity estimation in the beginning of the flight, while we flying velocity slowly drifting to some value (eg. +300 or -300), and when we enable ALTHOLD - estimated velocity has already a large drift, and I-term alone need some time to compensate this large drift. So instead we just subtract this large drift that gathered from the beginning of the flight. The next drifts are not so large (they smoothly accumulated each tick) and I term is compensating it well.

Reply:
You can model accelerometer offset and use Kalman to estimate it. It will be more complex in code, but simpler formally. Kalman can be used for optimal sensor fusion AND parameter estimation.

You are taking noise too strictly. All unmodeled/not measured effects are eventually equivalent to noise (at least as far as final PID controller is concerned).

Velocity after integration will be noisy to. (Leaky) integrator is  low-pass filter (https://dsp.stackexchange.com/questions/3179/is-a-leaky-integrator-the-same-thing-as-a-low-pass-filter), so it will only attenuate noise.

Velocity drift is direct result of accelerometer offset integration. With current implementation, velocity offset will eventually reach reach state where `velocityEstimationAccel * (1 - 0.999) == measuredAccel * 0.01f`. Now you simply subtract this offset (sampling it once), but you can use methods that will reduce (= estimate) it continuously.

`startVelocityEstimationAccel` - that's what I'm talking about - controller can't handle imperfect data, so you hack it to make error smaller. It does work, but it is not that great way to do it.
For example, can try to simulate ALTHOLD enable when going fast toward ground? IMO there will be considerable overshoot (PID I has to compensate sampled velocity)

Reply:
> can try to simulate ALTHOLD enable when going fast toward ground

I already mention this earlier:

> The only real improvement that probably should be done - using an additional source of velocity data to handle cases when you flying really fast towards ground (or climbing very fast).

Are you sure that Kalman is capable to detect a dynamic drift of accelerometer while quad is flying? How can it knows if it is constant drift or if user just constantly accelerating? It looks like physically impossible for me.

Reply:
You can't sustain vertical speed/acceleration for long time, at least under most conditions. You can assume that total altitude is bounded and then integral of velocity and (double integral of) acceleration must be close zero. Then speed s limited too, giving another constrain on acceleration.
So you have error signal than can be used to estimate drift. "All that is needed" is to use this information to get estimate of system state. 

Kalman is calculating optimal filter filter gains given system model and magnitude of noise sources. You can model gyro offset (as internal variable with it's associated noise), assume that height is constant (and actual height is now noise in this constant signal) and find optimal solution (optimal if all Kalman expectations are satisfied).

Or you can use other methodology to design control system - for example using stabilizing PID controllers or low-pass filters. Resulting algorithm will be very similar or equivalent.

As for `startVelocityEstimationAccel` : you can use low-pass filter on `velocityEstimationAccel`. Sampling this filtered value will be better than latching some random noise (with quite short time constant) or flight maneuver (longer filtering needed). [BTW: With long-enough filter, it will handle 'flight-into-ground' much better, velocity PID will start driving estimated velocity to zero immediately.]
But using above assumptions (flight envelope) to adjust estimated velocity is IMO cleaner and simpler.

Reply:
Sorry, i lost a thread a little bit.  What problem are you trying to solve?
1) Do you suggest to implement a kalman filter just to get rid of `startVelocityEstimationAccel`? 
2) Do you state that using kalman filter alone (without external altitude measurements with BARO or GPS) will solve the issue with "enter ALTHOLD while flying really fast towards ground"?

Reply:
1) Fixed case `enter ALTHOLD while flying really fast towards ground`.

Currently system slightly overreacts, I'm not sure how to make a good transition reaction - PID's that good for normal fly ("locked" stabilization) are not so good for "transition". In normal ALTHOLD flight speed is <5m/s and no fast accelerations. In transition - speed can be 30m/s and acceleration can be really high.
Also - in transition the initial thrust vector can be directed anywhere, it needs some time for ANGL to turn quad, and during this time PIDs increasing their reaction.

2) `startVelocityEstimationAccel` - left as is, since it's the simple and robust way to make system work.

Reply:
ok I'll give it a try and give you a review

Reply:
Wait, I haven't updated binaries yet.

Reply:
> Do you suggest to implement a kalman filter just to get rid of startVelocityEstimationAccel?

No, not at this point. I used Kalman as example how to do sensor fusion in mathematically sound way and as an example why velocity estimation is possible,

> Do you state that using kalman filter alone (without external altitude measurements with BARO or GPS) will solve the issue with "enter ALTHOLD while flying really fast towards ground"?

Yes, assuming bounds on flight envelope . But sing height input in estimate will greatly improve it (you can set higher gain in estimate controller).

> UPD. I can't use low-pass filter on velocityEstimationAccel - as I mentioned earlier - I don't know the exact values, and they differs from time-to-time. It's very unreliable solution. If I'll make a mistake and take too small values - the system will be completely broken. I can tune values for my quad, but at another quad values will differ, and users will get an oscillations out of control.

Estimation gains are dependent on noise (taking 'noise' as all unmeasured uncertainties). There will be always compromise between conservatism and speed of response. 

Did you try using PI controller, with speed as input and zero setpoint, with output added to accelerometer data?  P gain will replace current 'leak' (.999) and I term must be quite small, so that it tracks ACC offset only.


> Currently system slightly overreacts

Using some form of I-term windup prevention may help.

Reply:
> Using some form of I-term windup prevention may help.

Currently added blockbox logs and trying to find the reason. I will check how I-term behave in this situation and add temporary limits for transition period, if there is an issue with I-term.

My drone doesn't like this experiments. The last time I entered alt-hold flying 100kmh towards ground it just died. Looks like burned ESC.

Reply:
Can you post blackbox log the image is based on?

It is possible that leaky integrator is causing it - it does reduce velocity when gaining altitude, so there will be some overshoot in opposite direction when going down. Time constant is now proximately 10s, which may be a bit low. But increasing it will lead probably to other problems.

I'll look into it ...

Reply:
You need a complimentary filter to fuse some barometer and gps data into the gps velocity prediction, along with removing the leaky integral on the velocity. The complimentary filter will use the baro/gps to offer long term correction, and will use the acc for the instantaneous changes. I'd also look at better acc filtering to reduce the noise and delay that it currently has. Putting the rpm filter notches and maybe dynamic notches on the acc could likely reduce a lot of the filtering need that we have.

Reply:
Definitely something wrong with acceleration / velocity constant. Here is a baro-based velocity estimation and acc-based velocity estimation. They have difference x10 difference:
https://ibb.co/hK6LJKm

Reply:
Investigated, unfortunately there is no easy solution here. It's a regular accelerometer error.
Will be fixed later, in #11368 - probably will be added on-flight calibration and complementary filter.
Not a critical issue, it handles this case ok, after a small overreaction it stabilizing.